File: Beecrowd_Parte5/Um_Desafio_Simples_3097.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.StringTokenizer;

class FenwickTree {
    private int[] tree;
    private int maxVal;

    public FenwickTree(int maxSize) {
        this.tree = new int[maxSize + 1];
        this.maxVal = maxSize;
    }

    public void add(int idx, int value) {
        while (idx <= maxVal) {
            tree[idx] += value;
            idx += idx & -idx;
        }
    }

    public int sum(int idx) {
        int sum = 0;
        while (idx > 0) {
            sum += tree[idx];
            idx -= idx & -idx;
        }
        return sum;
    }

    public int rangeSum(int left, int right) {
        return sum(right) - sum(left - 1);
    }
}

public class Um_Desafio_Simples_3097 {

    public static void main(String[] args) throws IOException {
        BufferedReader reader = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer tokenizer = new StringTokenizer(reader.readLine());
        int N = Integer.parseInt(tokenizer.nextToken());
        int Q = Integer.parseInt(tokenizer.nextToken());

        int[] array = new int[N + 1];
        tokenizer = new StringTokenizer(reader.readLine());
        for (int i = 1; i <= N; i++) {
            array[i] = Integer.parseInt(tokenizer.nextToken());
        }

        FenwickTree fenwickTree = new FenwickTree(100000);
        int[] count = new int[N + 1];

        for (int i = 1; i <= N; i++) {
            fenwickTree.add(array[i], 1);
            count[i] = 0;
        }

        StringBuilder output = new StringBuilder();
        for (int q = 0; q < Q; q++) {
            tokenizer = new StringTokenizer(reader.readLine());
            int type = Integer.parseInt(tokenizer.nextToken());
            
            if (type == 1) {
                // Operação 1: Alterar o valor na posição A para Y
                int A = Integer.parseInt(tokenizer.nextToken());
                int Y = Integer.parseInt(tokenizer.nextToken());
                
                int currentVal = array[A];
                if (currentVal != Y) {
                    fenwickTree.add(currentVal, -1);
                    fenwickTree.add(Y, 1);
                    array[A] = Y;
                }
                
            } else if (type == 2) {
                // Operação 2: Consultar quantos números maiores que X existem no intervalo [A, B]
                int X = Integer.parseInt(tokenizer.nextToken());
                int A = Integer.parseInt(tokenizer.nextToken());
                int B = Integer.parseInt(tokenizer.nextToken());
                
                int greaterCount = fenwickTree.rangeSum(X + 1, 100000);
                output.append(greaterCount).append("\n");
            }
        }

        System.out.print(output);
    }
}

File: Beecrowd_Parte5/Um_Desafio_Simples_3097.c
#include <stdio.h>
#include <stdlib.h>

#define MAX_N 100000
#define MAX_Q 100000
#define MAX_VALUE 100000

typedef struct {
    int tree[MAX_VALUE + 1];
    int maxVal;
} FenwickTree;

void initFenwickTree(FenwickTree *ft, int maxSize) {
    ft->maxVal = maxSize;
    for (int i = 0; i <= maxSize; ++i) {
        ft->tree[i] = 0;
    }
}

void add(FenwickTree *ft, int idx, int value) {
    while (idx <= ft->maxVal) {
        ft->tree[idx] += value;
        idx += idx & (-idx);
    }
}

int sum(FenwickTree *ft, int idx) {
    int sum = 0;
    while (idx > 0) {
        sum += ft->tree[idx];
        idx -= idx & (-idx);
    }
    return sum;
}

int rangeSum(FenwickTree *ft, int left, int right) {
    return sum(ft, right) - sum(ft, left - 1);
}

int main() {
    int N, Q;
    int array[MAX_N + 1];
    FenwickTree ft;
    
    while (1) {
        scanf("%d %d", &N, &Q);
        if (N == 0 && Q == 0) break;
        
        // Read array elements
        for (int i = 1; i <= N; ++i) {
            scanf("%d", &array[i]);
        }
        
        // Initialize Fenwick Tree
        initFenwickTree(&ft, MAX_VALUE);
        
        // Build initial Fenwick Tree and count array
        for (int i = 1; i <= N; ++i) {
            add(&ft, array[i], 1);
        }
        
        // Process queries
        for (int q = 0; q < Q; ++q) {
            int type;
            scanf("%d", &type);
            
            if (type == 1) {
                // Operation 1: Update element at position A to Y
                int A, Y;
                scanf("%d %d", &A, &Y);
                
                int currentVal = array[A];
                if (currentVal != Y) {
                    add(&ft, currentVal, -1);
                    add(&ft, Y, 1);
                    array[A] = Y;
                }
                
            } else if (type == 2) {
                // Operation 2: Query how many numbers greater than X in range [A, B]
                int X, A, B;
                scanf("%d %d %d", &X, &A, &B);
                
                int greaterCount = rangeSum(&ft, X + 1, MAX_VALUE);
                printf("%d\n", greaterCount);
            }
        }
        
        // Print a blank line between test cases
        printf("\n");
    }
    
    return 0;
}
